fud2/fud-core/src/ast_converter.rs

```diff
@@ -0,0 +1,94 @@
+use cranelift_entity::PrimaryMap;
+use std::{collections::HashMap, ops};
+
+use crate::{
+    exec::{IO, OpRef, Operation},
+    flang::ast::{
+        Assignment, AssignmentList, Op, Visitable, Visitor, VisitorResult,
+    },
+};
+
+pub fn steps_to_ast(
+    plan: &Vec<(OpRef, Vec<IO>, Vec<IO>)>,
+    ops: &PrimaryMap<OpRef, Operation>,
+) -> AssignmentList {
+    let mut ast = AssignmentList { assigns: vec![] };
+    for step in plan {
+        let vars = step
+            .1
+            .iter()
+            .map(|v| match v {
+                IO::StdIO(utf8_path_buf) => utf8_path_buf,
+                IO::File(utf8_path_buf) => utf8_path_buf,
+            })
+            .cloned()
+            .collect();
+        let args = step
+            .2
+            .iter()
+            .map(|v| match v {
+                IO::StdIO(utf8_path_buf) => utf8_path_buf,
+                IO::File(utf8_path_buf) => utf8_path_buf,
+            })
+            .cloned()
+            .collect();
+
+        let fun = Op {
+            name: ops[step.0].name.clone(),
+            args,
+        };
+
+        let assignment = Assignment { vars, value: fun };
+        ast.assigns.push(assignment);
+    }
+
+    ast
+}
+
+/// A struct to convert a flang AST into the steps of a `Plan`.
+struct ASTToStepList {
+    step_list: Vec<(OpRef, Vec<IO>, Vec<IO>)>,
+    name_to_op_ref: HashMap<String, OpRef>,
+}
+
+impl ASTToStepList {
+    fn from_ops(ops: &PrimaryMap<OpRef, Operation>) -> Self {
+        let name_to_op_ref =
+            ops.iter().map(|(k, v)| (v.name.clone(), k)).collect();
+        ASTToStepList {
+            step_list: vec![],
+            name_to_op_ref,
+        }
+    }
+
+    fn step_list_from_ast(
+        &mut self,
+        ast: &AssignmentList,
+    ) -> Vec<(OpRef, Vec<IO>, Vec<IO>)> {
+        self.step_list = vec![];
+        let _ = ast.visit(self);
+        self.step_list.clone()
```

Comment:
Feel free to ignore this suggestion, but one Rusty approach I've used in similar situations, to avoid `clone`ing a big vector like this, is to make the `ASTToStepList` "disposable," as in the builder pattern. So this function would have a type signature like this:

```rs
    fn step_list_from_ast(
        self,
        ast: &AssignmentList,
    ) -> Vec<(OpRef, Vec<IO>, Vec<IO>)> {
```

That is, we take plain old `self` instead of `&mut self`, so this function takes ownership of and consumes `self`. Then the borrow checker will let us return `self.step_list` directly instead of `self.step_list.clone()`.